src/cpp/statistics/fastdds/domain/DomainParticipantImpl.cpp

```diff
@@ -0,0 +1,407 @@
+        return_code = find_or_create_topic_and_type(topic, topic_name, sample_identity_count_type);
+    }
+    else if (PHYSICAL_DATA_TOPIC == topic_name)
+    {
+        eprosima::fastdds::dds::TypeSupport physical_data_type(new PhysicalDataPubSubType);
+        return_code = find_or_create_topic_and_type(topic, topic_name, physical_data_type);
+    }
+    return return_code;
+}
+
+bool DomainParticipantImpl::find_or_create_topic_and_type(
+        eprosima::fastdds::dds::Topic** topic,
+        const std::string& topic_name,
+        const eprosima::fastdds::dds::TypeSupport& type) noexcept
+{
+    // Find if the topic has been already created and if the associated type is correct
+    eprosima::fastdds::dds::TopicDescription* topic_desc = lookup_topicdescription(topic_name);
+    if (nullptr != topic_desc)
+    {
+        if (topic_desc->get_type_name() != type->getName())
+        {
+            logError(STATISTICS_DOMAIN_PARTICIPANT, topic_name << " is not using expected type " << type->getName() <<
+                    " and is using instead type " << topic_desc->get_type_name());
+            return false;
+        }
+        else
+        {
+            // TODO(jlbueno) This casting should be checked after other TopicDescription implementations are
+            // included: ContentFilteredTopic, MultiTopic.
+            *topic = dynamic_cast<eprosima::fastdds::dds::Topic*>(topic_desc);
+        }
+    }
+    else
+    {
+        if (ReturnCode_t::RETCODE_PRECONDITION_NOT_MET == register_type(type, type->getName()))
+        {
+            // No log because it is already logged within register_type
+            return false;
+        }
+        // Create topic. No need to check return pointer. It fails if the topic already exists, if the QoS is
+        // inconsistent or if the type is not registered.
+        *topic = create_topic(topic_name, type->getName(), eprosima::fastdds::dds::TOPIC_QOS_DEFAULT);
+    }
+    assert(nullptr != *topic);
+    return true;
+}
+
+bool DomainParticipantImpl::delete_topic_and_type(
+        const std::string& topic_name) noexcept
+{
+    eprosima::fastdds::dds::TopicDescription* topic_desc = lookup_topicdescription(topic_name);
+    assert(nullptr != topic_desc);
+    eprosima::fastdds::dds::Topic* topic = dynamic_cast<eprosima::fastdds::dds::Topic*>(topic_desc);
+    std::string type_name = topic->get_type_name();
+    // delete_topic can fail if the topic is referenced by any other entity. This case could happen even if
+    // it should not. It also fails if topic is a nullptr (dynamic_cast failure).
+    if (ReturnCode_t::RETCODE_OK != delete_topic(topic))
+    {
+        return false;
+    }
+    // unregister_type failures are of no concern here. It will fail if the type is still in use (something
+    // expected) and if the type_name is empty (which is not going to happen).
+    unregister_type(type_name);
+    return true;
```

Comment:
Our usual failure policy is to try to fulfill as many of the method's tasks as possible. Shouldn't we try to unregister the type even if the topic deletion fails?

Reply:
The type is being used by the topic, so if the topic deletion fails, then `unregister_type` would also fail.